### PR TITLE
[KOGITO-5622] Implement Generator.isEmpty() to skip generation when no resources are available

### PR DESCRIPTION
**Pull request by danielezonca**
https://issues.redhat.com/browse/KOGITO-5622

## Details
Kogito has a compile time execution (codegen) and a runtime execution (engine execution).
Kogito integration with Quarkus and Springboot is done differently:
- in Quarkus there are multiple extensions that implicitly brings in both the compile time (aka `deployment` using Quarkus naming) and the runtime for the selected engine (DMN, BPMN etc)
- in Springboot we have a single `kogito-maven-plugin` with all compile time code for all the possible engines even if the runtime part is not available

Quarkus behaviour is already correct while with SpringBoot there is a side effect: `kogito-maven-plugin` is performing some generation even for engines that are not available that can lead to compilation errors. This was not really possible until now but the new specific springboot starters made this problem now a blocker. 

This PR introduce a simple `isEmpty` check so that a generator that has no resources to consume is automatically disabled.

There is a corner case that it is still not properly addressed by this ticket: what if for example a user includes only processes but then in the resource folder provides also a DRL file? At the moment the `isEmpty` check will pass so the rule generator will execute.
We might consider to add a check at effective pom level to see if a runtime feature is available or we might decide to add a "marker" class to check to see if the project is properly configured. A similar check might require to have a project classLoader (aka a classloader with only runtime classpath) that is not available in Quarkus context (see [KOGITO-4270](https://issues.redhat.com/browse/KOGITO-4270))

Please make sure that your PR meets the following requirements:

- [x] You have read the [contributors guide](CONTRIBUTING.md)
- [x] Your code is properly formatted according to [this configuration](https://github.com/kiegroup/kogito-runtimes/tree/main/kogito-build/kogito-ide-config)
- [x] Pull Request title is properly formatted: `KOGITO-XYZ Subject`
- [x] Pull Request title contains the target branch if not targeting main: `[0.9.x] KOGITO-XYZ Subject`
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains link to any dependent or related Pull Request
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* <b>Run all builds</b>  
  Please add comment: <b>Jenkins retest this</b>

* <b>Run (or rerun) specific test(s)</b>  
  Please add comment: <b>Jenkins (re)run [runtimes|optaplanner|apps|examples] tests</b>
 
* <b>Quarkus LTS checks</b>  
  Please add comment: <b>Jenkins run LTS</b>

* <b>Run (or rerun) LTS specific test(s)</b>  
  Please add comment: <b>Jenkins (re)run [runtimes|optaplanner|apps|examples] LTS</b>

* <b>Native checks</b>  
  Please add comment: <b>Jenkins run native</b>

* <b>Run (or rerun) native specific test(s)</b>  
  Please add comment: <b>Jenkins (re)run [runtimes|optaplanner|apps|examples] native</b>

* <b>Full Kogito testing</b> (with cloud images and operator BDD testing)  
  Please add comment: <b>Jenkins run BDD</b>  
  <b>This check should be used only if a big change is done as it takes time to run, need resources and one full BDD tests check can be done at a time ...</b>
</details>

### CODE REVIEW

**Comment by ricardozanini** [comment]
Do we have a follow-up JIRA for the corner case you just described? Maybe we could even have our plugin to verify the POM dependencies and the resources presented in the project, failing the compilation if the generator is not present in the classpath.

**Review by ricardozanini — COMMENTED** [review]
Seems pretty straightforward. I'll test it on my branch with some examples that failed in the last run.

File: kogito-codegen-modules/kogito-codegen-processes/src/main/java/org/kie/kogito/codegen/process/ProcessCodegen.java [hunk 1]
```diff
@@ -266,19 +266,8 @@ public ProcessCodegen(KogitoBuildContext context, Collection<? extends Process>
         }
     }
 
-    public static String defaultWorkItemHandlerConfigClass(String packageName) {
```
[the remark this answers is not in the document]
Comment (danielezonca) [reply kogito-codegen-modules/kogito-codegen-processes/src/main/java/org/kie/kogito/codegen/process/ProcessCodegen.java:269]: Yes it is unrelated to this PR but the IDE marks them as unused so I did this small cleanup

**Review by cristianonicolai — APPROVED** [review]

File: kogito-codegen-modules/kogito-codegen-processes/src/test/java/org/kie/kogito/codegen/process/persistence/FileSystemPersistenceGeneratorTest.java [hunk 2]
```diff
@@ -62,30 +64,33 @@ void test() {
...
         assertThat(generatedFiles.stream().filter(gf -> gf.type().equals(ProtoGenerator.PROTO_TYPE)).count()).isEqualTo(2);
         assertThat(generatedFiles.stream().filter(gf -> gf.type().equals(ProtoGenerator.PROTO_TYPE) && gf.relativePath().endsWith(".json")).count()).isEqualTo(1);
-        assertThat(generatedFiles).hasSize(15);
+        int expectedNumberOfFiles = context.hasRESTForGenerator(persistenceGenerator) ? 15 : 14;
+        assertThat(generatedFiles).hasSize(expectedNumberOfFiles);
 
-        Optional<GeneratedFile> persistenceFactoryImpl = generatedFiles.stream()
-                .filter(gf -> gf.relativePath().equals("org/kie/kogito/persistence/KogitoProcessInstancesFactoryImpl.java"))
-                .findFirst();
+        if (context.hasRESTForGenerator(persistenceGenerator)) {
```
Comment (MarianMacik) [inline kogito-codegen-modules/kogito-codegen-processes/src/test/java/org/kie/kogito/codegen/process/persistence/FileSystemPersistenceGeneratorTest.java:70]: Persistence factory is not generated if REST generation is disabled?

**Review by MarianMacik — APPROVED** [review]
Looks good, just one comment.

File: kogito-codegen-modules/kogito-codegen-processes/src/test/java/org/kie/kogito/codegen/process/persistence/FileSystemPersistenceGeneratorTest.java (hunk 2 above)
Comment (danielezonca) [reply kogito-codegen-modules/kogito-codegen-processes/src/test/java/org/kie/kogito/codegen/process/persistence/FileSystemPersistenceGeneratorTest.java]: My bad, the condition should be `hasDI()`. At the moment persistence is not available in an environment without dependency injection, this is possible only in our tests but it is a branch that we still keep working.
Fixed

**Review by ricardozanini — APPROVED** [review]

**Comment by ricardozanini** [comment]
jenkins rerun examples tests
